lib/api/apiUtils/object/corsResponse.js

```diff
@@ -0,0 +1,127 @@
+    const allowedHeaders = ruleAllowedHeaders.map(allowedHeader =>
+        allowedHeader.toLowerCase());
+    if (!headers.every(header => _matchesOneOf(allowedHeaders, header))) {
+        return false;
+    }
+    return true;
+}
+
+/** _findCorsRule - Return first matching rule in cors rules that permits
+*   CORS request
+* @param {object[]} rules - array of rules
+* @param {string} [rules.id] - optional id to identify rule
+* @param {string[]} rules[].allowedMethods - methods allowed for CORS
+* @param {string[]} rules[].allowedOrigins - origins allowed for CORS
+* @param {string[]} [rules[].allowedHeaders] - headers allowed in an
+*   OPTIONS request via the Access-Control-Request-Headers header
+* @param {number} [rules[].maxAgeSeconds] - seconds browsers should cache
+*   OPTIONS response
+* @param {string[]} [rules[].exposeHeaders] - headers to expose to external
+*   applications
+* @param {string} origin - origin of CORS request
+* @param {string} method - expected method of CORS request
+* @param {string[]} [headers] - expected headers of CORS request
+* @return {(null|object)} - matching rule if found; null if no match
+*/
+export function findCorsRule(rules, origin, method, headers) {
+    for (let i = 0; i < rules.length; i++) {
+        const rule = rules[i];
+        if (rule.allowedMethods.indexOf(method) === -1) {
+            continue;
+        } else if (!_matchesOneOf(rule.allowedOrigins, origin)) {
+            continue;
+        } else if (headers &&
+            !_headersMatchRule(headers, rule.allowedHeaders)) {
+            continue;
+        }
+        // if it passed all above checks, found matching rule
+        return rule;
+    }
+    // searched all rules and found no match
+    return null;
+}
+
+/** _gatherResHeaders - Collect headers to return in response
+* @param {object} rule - array of rules
+* @param {string} [rule.id] - optional id to identify rule
+* @param {string[]} rule[].allowedMethods - methods allowed for CORS
+* @param {string[]} rule[].allowedOrigins - origins allowed for CORS
+* @param {string[]} [rule[].allowedHeaders] - headers allowed in an
+*   OPTIONS request via the Access-Control-Request-Headers header
+* @param {number} [rule[].maxAgeSeconds] - seconds browsers should cache
+*   OPTIONS response
+* @param {string[]} [rule[].exposeHeaders] - headers to expose to external
+*   applications
+* @param {string} origin - origin of CORS request
+* @param {string} method - expected method of CORS request
+* @param {string[]} [headers] - expected headers of CORS request
+* @return {object} resHeaders - headers to include in response
+*/
+export function gatherCorsResHeaders(rule, origin, method, headers) {
+    const resHeaders = {
+        'access-control-max-age': rule.maxAgeSeconds,
+        'access-control-allow-methods': rule.allowedMethods.join(', '),
+        'content-length': '0',
+        'date': new Date().toUTCString(),
+        'vary':
```

Comment:
As far as I can tell, AWS always includes the same three values: `Origin, Access-Control-Request-Method, Access-Control-Request-Headers` as long as the response is not 403 Forbidden. This is even if rule Origin is `'*'` and even if the request includes no values for Access-Control-Request-Headers.

The only thing MDN mentions regarding the Vary header for CORS:

> If the server specifies an origin host rather than "*", then it must also include Origin in the Vary response header to indicate to clients that server responses will differ based on the value of the Origin request header.

But it does seem to make sense since these are the three headers AWS uses to determine whether to send a 200 OK response or 403 AccessForbidden response to the preflight requests.